File: frontend/src/utils/profileValidation.ts
import { getBasicErrorFromUnknown } from '@/api/client'
import type { BasicError } from '@/models/apiError'

// eslint-disable-next-line no-control-regex -- reject control chars in profile text fields
const CONTROL_CHAR = /[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]/

export const NICKNAME_MIN = 3
export const NICKNAME_MAX = 16
export const REAL_NAME_MIN = 2
export const REAL_NAME_MAX = 64

export function normalizeMobile(value: string): string | null {
  const trimmed = value.trim()
  if (!trimmed) return null
  const stripped = trimmed.replace(/[\s\-()]/g, '')
  if (!stripped) return null
  const plusCount = (stripped.match(/\+/g) ?? []).length
  if (plusCount > 1 || (stripped.includes('+') && !stripped.startsWith('+'))) {
    return null
  }
  if (!/^\+?[0-9]{7,20}$/.test(stripped)) {
    return null
  }
  return stripped
}

export function validateNickname(
  value: string,
  t: (key: string) => string,
): string | null {
  const v = value.trim()
  if (!v) return t('nicknameMinLength')
  if (v.length < NICKNAME_MIN) return t('nicknameMinLength')
  if (v.length > NICKNAME_MAX) return t('nicknameMaxLength')
  if (CONTROL_CHAR.test(v)) return t('nicknameInvalid')
  return null
}

export function validateRealName(
  value: string,
  t: (key: string) => string,
): string | null {
  const v = value.trim()
  if (!v) return null
  if (v.length < REAL_NAME_MIN) return t('realNameMinLength')
  if (v.length > REAL_NAME_MAX) return t('realNameMaxLength')
  if (CONTROL_CHAR.test(v)) return t('realNameInvalid')
  return null
}

export function validateMobile(
  value: string,
  t: (key: string) => string,
): string | null {
  const v = value.trim()
  if (!v) return null
  const normalized = normalizeMobile(v)
  if (normalized === null) return t('mobileInvalid')
  return null
}

export function mapProfileApiMessage(msg: string, t: (key: string) => string): string {
  if (msg === 'duplicate mobile') return t('mobileDuplicate')
  return msg
}

export function profileErrorFromUnknown(err: unknown, t: (key: string) => string): BasicError {
  const basic = getBasicErrorFromUnknown(err)
  return {
    msg: mapProfileApiMessage(basic?.msg ?? 'error', t),
    detail: basic?.detail,
  }
}

File: utils/profile_validation.py
import re

from models import User

_CONTROL_CHAR_PATTERN = re.compile(r'[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]')
_MOBILE_STRIP_PATTERN = re.compile(r'[\s\-()]')
_MOBILE_VALID_PATTERN = re.compile(r'^\+?[0-9]{7,20}$')

NICKNAME_MIN_LENGTH = 3
NICKNAME_MAX_LENGTH = 16
REAL_NAME_MIN_LENGTH = 2
REAL_NAME_MAX_LENGTH = 64


class ProfileValidationError(ValueError):
    pass


def normalize_optional_text(value: str | None) -> str | None:
    if value is None:
        return None
    if not isinstance(value, str):
        raise ProfileValidationError('value must be a string')
    stripped = value.strip()
    return stripped if stripped else None


def _reject_control_chars(value: str) -> None:
    if _CONTROL_CHAR_PATTERN.search(value):
        raise ProfileValidationError('invalid characters')


def validate_nickname(value: str | None) -> str | None:
    normalized = normalize_optional_text(value)
    if normalized is None:
        return None
    _reject_control_chars(normalized)
    if len(normalized) < NICKNAME_MIN_LENGTH:
        raise ProfileValidationError('invalid nickname format')
    if len(normalized) > NICKNAME_MAX_LENGTH:
        raise ProfileValidationError('invalid nickname format')
    return normalized


def validate_real_name(value: str | None) -> str | None:
    normalized = normalize_optional_text(value)
    if normalized is None:
        return None
    _reject_control_chars(normalized)
    if len(normalized) < REAL_NAME_MIN_LENGTH:
        raise ProfileValidationError('invalid real name format')
    if len(normalized) > REAL_NAME_MAX_LENGTH:
        raise ProfileValidationError('invalid real name format')
    return normalized


def normalize_mobile(value: str | None) -> str | None:
    normalized = normalize_optional_text(value)
    if normalized is None:
        return None
    stripped = _MOBILE_STRIP_PATTERN.sub('', normalized)
    if not stripped:
        return None
    if stripped.count('+') > 1 or (stripped.startswith('+') and '+' in stripped[1:]):
        raise ProfileValidationError('invalid mobile format')
    if '+' in stripped and not stripped.startswith('+'):
        raise ProfileValidationError('invalid mobile format')
    if not _MOBILE_VALID_PATTERN.match(stripped):
        raise ProfileValidationError('invalid mobile format')
    return stripped


def validate_mobile(value: str | None) -> str | None:
    return normalize_mobile(value)


def display_name(user: User) -> str:
    for attr in ('nickname', 'real_name', 'name'):
        val = getattr(user, attr, None)
        if isinstance(val, str) and val.strip():
            return val.strip()
    return user.name
